server/datastore/mysql/software_titles.go

```diff
@@ -51,6 +51,19 @@ func (ds *Datastore) ListSoftwareTitles(
 	ctx context.Context,
 	opt fleet.SoftwareTitleListOptions,
 ) ([]fleet.SoftwareTitle, int, *fleet.PaginationMetadata, error) {
+	if opt.ListOptions.After != "" {
+		return nil, 0, nil, fleet.NewInvalidArgumentError("after", "not supported for software titles")
+	}
+
+	if len(strings.Split(opt.ListOptions.OrderKey, ",")) > 1 {
+		return nil, 0, nil, fleet.NewInvalidArgumentError("order_key", "multicolumn order key not supported for software titles")
+	}
+
+	if opt.ListOptions.OrderKey == "" {
+		opt.ListOptions.OrderKey = "hosts_count"
+		opt.ListOptions.OrderDirection = fleet.OrderDescending
+	}
+
```

Comment:
IIRC most of our validation lives at the service level

Reply:
Yes, that's right. In general, we do validations on the incoming query params at the service layer. In this case, I wanted to keep these validations at the datastore level because they are closely related to a specific problem in the SQL layer (namely the `sanitizeColumn` function used by `appendListOptionsWIthCursorSQL` escapes `,` and generally can't handle multicolumn sort). This way we can test them directly in the SQL layer and we'll just remove the validations from the SQL layer once the SQL issue is fixed. Roberto and I discussed this last week and we decided to do a targeted fix for the software titles feature and plan to address the bigger issue in a follow up ticket.

Reply:
ah makes sense, that will also protect other functions from using invalid ListOptions when calling this method